File: C++/similar-string-groups.cpp
// Time:  O(n^2 * l), l is the average length of words
// Space: O(n)

class Solution {
public:
    int numSimilarGroups(vector<string>& A) {
        UnionFind union_find(A.size());
        for (int i = 0; i < A.size(); ++i) {
            for (int j = 0; j < i; ++j) {
                if (isSimilar(A[i], A[j])) {
                    union_find.union_set(i, j);
                }
            }
        }
        return union_find.size();
    }
    
private:
    bool isSimilar(const string& a, const string& b) {
        int diff = 0;
        for (int i = 0; i < a.length(); ++i) {
            if (a[i] != b[i]) {
                if (++diff > 2) {
                    return false;
                }
            }
        }
        return diff == 2;
    }
    
    class UnionFind {
        public:
            UnionFind(const int n) : set_(n), size_(n) {
                iota(set_.begin(), set_.end(), 0);
            }

            int find_set(const int x) {
               if (set_[x] != x) {
                   set_[x] = find_set(set_[x]);  // Path compression.
               }
               return set_[x];
            }

            bool union_set(const int x, const int y) {
                int x_root = find_set(x), y_root = find_set(y);
                if (x_root == y_root) {
                    return false;
                }
                set_[min(x_root, y_root)] = max(x_root, y_root);
                --size_;
                return true;
            }
        
            int size() const {
                return size_;
            }

        private:
            vector<int> set_;
            int size_;
    };
};

// Time:  O(n^2 * l) ~ O(n * l^4)
// Space: O(n) ~ O(n * l^3)
class Solution_MLE {
public:
    int numSimilarGroups(vector<string>& A) {
        const int N = A.size(), L = A[0].length();
        UnionFind union_find(A.size());
        if (N < L*L) {
            for (int i = 0; i < N; ++i) {
                for (int j = 0; j < i; ++j) {
                    if (isSimilar(A[i], A[j])) {
                        union_find.union_set(i, j);
                    }
                }
            }
        } else {
            unordered_map<string, vector<int>> buckets;
            unordered_set<string> lookup;
            for (int i = 0; i < A.size(); ++i) {
                auto word = A[i];
                if (!lookup.count(word)) {
                    buckets[word].emplace_back(i);
                    lookup.emplace(word);
                }
                for (int j1 = 0; j1 < L; ++j1) {
                    for (int j2 = 0; j2 < j1; ++j2) {
                        swap(word[j1], word[j2]);
                        buckets[word].emplace_back(i);
                        swap(word[j1], word[j2]);
                    }
                }
            }
            for (const auto& word : A) {
                for (int i = 0; i < buckets[word].size(); ++i) {
                    for (int j = 0; j < i; ++j) {
                        union_find.union_set(buckets[word][i], buckets[word][j]);
                    }
                }
            }
        }
        return union_find.size();
    }
    
private:
    bool isSimilar(const string& a, const string& b) {
        int diff = 0;
        for (int i = 0; i < a.length(); ++i) {
            if (a[i] != b[i]) {
                if (++diff > 2) {
                    return false;
                }
            }
        }
        return diff == 2;
    }
    
    class UnionFind {
        public:
            UnionFind(const int n) : set_(n), size_(n) {
                iota(set_.begin(), set_.end(), 0);
            }

            int find_set(const int x) {
               if (set_[x] != x) {
                   set_[x] = find_set(set_[x]);  // Path compression.
               }
               return set_[x];
            }

            bool union_set(const int x, const int y) {
                int x_root = find_set(x), y_root = find_set(y);
                if (x_root == y_root) {
                    return false;
                }
                set_[min(x_root, y_root)] = max(x_root, y_root);
                --size_;
                return true;
            }
        
            int size() const {
                return size_;
            }

        private:
            vector<int> set_;
            int size_;
    };
};

File: Python/similar-string-groups.py
# Time:  O(n^2 * l) ~ O(n * l^4)
# Space: O(n) ~ O(n * l^3)

import collections
import itertools


class UnionFind(object):
    def __init__(self, n):
        self.set = range(n)
        self.__size = n

    def find_set(self, x):
        if self.set[x] != x:
            self.set[x] = self.find_set(self.set[x])  # path compression.
        return self.set[x]

    def union_set(self, x, y):
        x_root, y_root = map(self.find_set, (x, y))
        if x_root == y_root:
            return False
        self.set[min(x_root, y_root)] = max(x_root, y_root)
        self.__size -= 1
        return True

    def size(self):
        return self.__size


class Solution(object):
    def numSimilarGroups(self, A):
        def isSimilar(a, b):
            diff = 0
            for x, y in itertools.izip(a, b):
                if x != y:
                    diff += 1
                    if diff > 2:
                        return False
            return diff == 2

        N, L = len(A), len(A[0])
        union_find = UnionFind(N)
        if N < L*L:
            for (i1, word1), (i2, word2) in \
                    itertools.combinations(enumerate(A), 2):
                if isSimilar(word1, word2):
                    union_find.union_set(i1, i2)
        else:
            buckets = collections.defaultdict(list)
            lookup = set()
            for i in xrange(len(A)):
                word = list(A[i])
                if A[i] not in lookup:
                    buckets[A[i]].append(i)
                    lookup.add(A[i])
                for j1, j2 in itertools.combinations(xrange(L), 2):
                    word[j1], word[j2] = word[j2], word[j1]
                    buckets["".join(word)].append(i)
                    word[j1], word[j2] = word[j2], word[j1]
            for word in A:  # Time:  O(n * l^4)
                for i1, i2 in itertools.combinations(buckets[word], 2):
                    union_find.union_set(i1, i2)
        return union_find.size()

